modules/stochastic_tools/include/utils/Calculators.h

```diff
@@ -125,7 +212,525 @@ class L2Norm : public Calculator<InType, OutType>
  * Explicit instantiations in C file.
  */
 template <typename InType = std::vector<Real>, typename OutType = Real>
-std::unique_ptr<const Calculator<InType, OutType>>
-makeCalculator(const MooseEnumItem & item, const libMesh::ParallelObject & other);
+std::unique_ptr<Calculator<InType, OutType>> makeCalculator(const MooseEnumItem & item,
+                                                            const libMesh::ParallelObject & other);
+
+/**
+ * This class is used as a general interface for doing arithmetic needed for Calculators.
+ * The idea is that instead of redefining each calculator method for new in-out data types,
+ * one can just redefine these simple operations. The operations defined here are just
+ * for scalar value types like Real and int.
+ *
+ * @tparam T1 the "in-value" type, this is the Intype::value_type in the Calculators
+ * @tparam T2 the "out-value" type, this is underlying type of this class and is what
+ *            is returned with the get() function. This is the OutType in the Calculators
+ */
+template <typename T1, typename T2>
+class CalculatorValue
+{
+public:
+  CalculatorValue() : _value() {}
+
+  /// Returns a reference to the value computed
+  const T2 & get() const { return _value; }
+
+  /**
+   * These are functions that do modifications on the value that do not depend
+   * on the in-type @tparam T1.
+   */
+  ///@{
+  /// Set the value to zero
+  void zero() { _value = T2(); }
+  /// Divide the value by a interger value
+  void divide(dof_id_type num) { _value /= num; }
+  /// Perfoming a exponential operation: _value = _value^p
+  void pow(int p) { _value = MathUtils::pow(_value, p); }
+  /// Square root of the value
+  void sqrt() { _value = std::sqrt(_value); }
+  /// Setting the value to the minimum of the data type
+  void min() { _value = std::numeric_limits<T2>::min(); }
+  /// Setting the value to the maximum of the data type
+  void max() { _value = std::numeric_limits<T2>::max(); }
+  ///@}
+
+  /**
+   * These are functions that modify the value with an in-type value
+   */
+  ///@{
+  /// _value += a
+  void add(const T1 & a) { _value += static_cast<T2>(a); }
+  /// _value += a^p
+  void addPow(const T1 & a, int p) { _value += MathUtils::pow(static_cast<T2>(a), p); }
+  /// _value = min(_value, a)
+  void min(const T1 & a) { _value = std::min(static_cast<T2>(a), _value); }
+  /// _value = max(_value, a)
+  void max(const T1 & a) { _value = std::max(static_cast<T2>(a), _value); }
+  ///@}
+
+  /**
+   * These are overloaded operators that modify the value based on out-type values
+   */
+  ///@{
+  CalculatorValue<T1, T2> & operator+=(const T2 & b)
+  {
+    _value += b;
```

Comment:
Could you utilize the above methods, so that these do not need to be defined in the specializations and work for all instantiated CalculatorValues?

Reply:
Also, what about `greaterThan` and `equal`?

Reply:
I added functions based on what was needed by the calculators. `a.greaterThan(b)` is the same as `b.lessThan(a)` and `a.equal(b)` is the same as `!(a.lessThan(b) || b.lessThan(a))`.

I'm starting to think there are too many subtleties in this implementation. You basically have to understand how the calculator is using these functions to define them correctly, which is what I was trying to avoid in the first place. I'm thinking of going back and doing specializations of the calculators themselves.